Data URI support, redirect checks and more

Bhpsngum opened the pull request:
(From issue #25)
## Data URI protocol (`data:`)
- Add basic support for data URI protocol
- Support base64 encoding, allowing ASCII space characters (`' '`, `'\t'`, `'\n'`, `'\v'`, `'\f'`, `'\r'`) in encoded + non-encoded form, and strict padding checks
- Failure of validation will throw/dispatch "Invalid URL" error
## Redirects
- Add support for HTTP codes 301 and 308
- Add new option `new XMLHttpRequest({ maxRedirects: 20 /* default is 20 */ })`
- Excessive redirects will throw/dispatch an error `Too many redirects`, and attempting to redirect to any non-secure protocols (other than `https` and `http`) or invalid URI throws/dispatches `Unsafe redirect` error.
## Local file system
- Add new option `new XMLHttpRequest({ allowFileSystemSources: true /* default is true */ })`
- Attempting to load from local filesystem resource while disabling this option throws/dispatches `Not allowed to load local resources: file://...` error.
## Origin
- Add new option `new XMLHttpRequest({ origin: "https://example.com" /* default is no origin */ })`
- Similar to calling xhr in the browser page, the origin will act as a base URL, if origin is blank or invalid, any URL passed through `xhr.open` will act as an absolute URL.
## Others
- Add `xhr.responseURL` to reflect XMLHttpRequest spec
- `xhr.statusText` now returns the correct status text, taken from `HttpIncomingMessage.statusMessage`
- `xhr.dispatchEvent` now dispatches additional object to event parameter `{ type: "event_type" }`, e.g `{ type: "loadend" }`
- Fix `xhr.getResponseHeader()` and `xhr.getAllResponseHeaders()` throwing errors when no responses or headers are present (for local resources)
## Codespace
- Changing from `Url.parse()` to new WHATWG-compatible `new URL()` with more safe restrictions
- Load optional options from default options to prevent parameters prototype pollution
- Destroy any redirect responses to prevent hanging
- SSL options are now separated from original option for better code control

Bhpsngum committed:
Data URI support, redirect checks and more

mjwwit commented:
Thanks for this! Would you be willing to add some tests for this functionality as well?

Bhpsngum commented:
sure, give me a bit

Bhpsngum committed:
add tests

Bhpsngum committed:
Fix one of the tests

Bhpsngum commented:
I added the tests, you can have a look around tho

Bhpsngum committed:
allowFileSystemSources --> allowFileSystemResources

Bhpsngum committed:
Add origin options

mjwwit commented on lib/XMLHttpRequest.js:
```diff
@@ -404,7 +413,13 @@ function XMLHttpRequest(opts) {
     var url;
     try {
-      url = new URL(settings.url);
+      if (settings.origin != null) {
```
Please use `!==`, as it is more accurate

Bhpsngum replied:
This part is intentional as it checks for both `null` and `undefined` (as both of them indicate that user does not specify any origin)

mjwwit replied:
In that case it's probably better to use `if (settings.origin) {`, as this will also filter `false` and other falsy values.

Bhpsngum replied:
That works as well, will do.

mjwwit commented on lib/XMLHttpRequest.js:
```diff
@@ -37,7 +36,34 @@ XMLHttpRequest.XMLHttpRequest = XMLHttpRequest;
-  opts = opts || {};
+  // defines a list of default options to prevent parameters pollution
+  var default_options = {
+    pfx: void 0,
+    key: void 0,
+    passphrase: void 0,
+    cert: void 0,
+    ca: void 0,
+    ciphers: void 0,
```
`undefined` would be just fine instead of `void 0`.

Bhpsngum replied:
sure i will change them to `undefined`

mjwwit commented on lib/XMLHttpRequest.js:
```diff
@@ -365,8 +410,21 @@ function XMLHttpRequest(opts) {
+      settings.url = url.href;
+    }
+    catch (e) {
+      self.handleError(new Error("ERR_INVALID_URL"));
```
Why not handle the original error `e` instead?

Bhpsngum replied:
browsers usually return "ERR_INVALID_URL" if the given url is invalid, specifying the error in parsing URL is not that necessary imo.

mjwwit replied:
This is actually the exact error that will be thrown, hence my comment. However, your error has an `"ERR_INVALID_URL"` `message`, while the actual error has it as a `code`.

Bhpsngum replied:
For URL constructor, it throws a message like "Invalid Base URL" or "Invalid URL", I could switch to throwing that message instead.

Yeah, this one could be helpful in case the user defines an invalid origin (so user may have more idea what's going wrong with their URL).

Bhpsngum replied:
```js
> try { new URL("").href } catch (e) {
    console.log(e.message);
}

< "Failed to construct 'URL': Invalid base URL"
```
However, this is the original message after it throws, should i just either keep the last part (Invalid URL) or just keep the whole message?

The implementation seem to be different in Node, as it only throws the latter, so i guess could keep everything.

mjwwit commented on lib/XMLHttpRequest.js:
```diff
@@ -365,8 +410,21 @@ function XMLHttpRequest(opts) {
 
-    var ssl = false, local = false;
-    var url = Url.parse(settings.url);
+    var ssl = false, local = false, datauri = false;
```
If we name these like the flags they are, determining their value type elsewhere will be much easier: `isSsl`, `isLocal`, `isDataUri`.

Bhpsngum replied:
the `ssl` and `local` are from old implementation, i just simply add `datauri` to match with them. However if you want i can refactor these.

Bhpsngum replied:
i refactored all codes and change to `isLocal`, `isDataUri` and `isSsl`

mjwwit commented on lib/XMLHttpRequest.js:
```diff
@@ -377,6 +435,9 @@ function XMLHttpRequest(opts) {
 
+      case 'data:':
+        datauri = true;
+      
```
I don't think multiple protocols can be valid at the same time, so a `break` is missing here

Bhpsngum replied:
for data URI i parse the URL inside `local` check like this:
```js
if (local) {
	if (datauri) {
		// parse Data URI
		return;
	}
	// parse file protocol
	return;
}
```

so a non-break here is intentional, as the next step it marks the `local` flag, and allows the data URI to be parsed like in the logic above

mjwwit commented on lib/XMLHttpRequest.js:
```diff
@@ -391,7 +452,61 @@ function XMLHttpRequest(opts) {
     if (local) {
+      if (datauri) try {
+        // Triming from original url string since 
+        var data = url.href.substr(5, settings.url.length);
```
`substr` is a deprecated legacy Internet Explorer String method, we should probably use either `substring` or `slice` here. Do note that both use an end index instead of the length as a second parameter.

Bhpsngum replied:
I use `substr` as i could see the functionality being used in other parts of the function. However again, i could change to `slice` or `substring`, need to check with the method supports wider range of node versions.

Bhpsngum replied:
I changed every usage of `substr` to `slice`

mjwwit commented on lib/XMLHttpRequest.js:
```diff
@@ -391,7 +452,61 @@ function XMLHttpRequest(opts) {
     if (local) {
+      if (datauri) try {
+        // Triming from original url string since 
+        var data = url.href.substr(5, settings.url.length);
+
+        // separator between header and actual data
+        var parts = data.split(",", 2);
+
+        if (parts.length < 2) throw "Invalid URL";
+
+        var base64 = false;
+        // check if header part has base64
+        var data_headers = parts[0].split(";");
+        
+        // ignore first element since it's MIME type
+        for (var i = 1; i < data_headers.length; ++i) {
+          if (data_headers[i].toLowerCase() === "base64") {
+            base64 = true;
+            break;
+          }
+        }
+
+        var responseData, inputData = decodeURIComponent(parts[1]);
+
+        if (base64) {
+          // remove any ASCII whitespaces
+          inputData = inputData.replace(/(\s|\t|\r|\n|\v|\f)+/g, "");
+          // check padding amount
+          let padding = inputData.match(/=*$/)[0];
+          if (padding.length + (inputData.length - padding.length) % 4 > 4) throw "invalid padding";
+          inputData = inputData.substr(0, inputData.length - padding.length);
+          responseData = Buffer.from(inputData, "base64");
+          if (responseData.toString("base64").replace(/=+$/, "") !== inputData) throw "malformed base64 encoding";
+        }
+        else {
+          responseData = Buffer.from(inputData);
+        }
```
Would it be more clear to extract this to a separate function? `toBufferFromDataUri` or something in those lines. I realize that is definitely not the trend in this project, but there's no reason to keep adding to the madness.

Bhpsngum replied:
sure, let me separate it

mjwwit commented on lib/XMLHttpRequest.js:
```diff
@@ -391,7 +452,61 @@ function XMLHttpRequest(opts) {
+        var data_headers = parts[0].split(";");
+        
+        // ignore first element since it's MIME type
+        for (var i = 1; i < data_headers.length; ++i) {
+          if (data_headers[i].toLowerCase() === "base64") {
+            base64 = true;
+            break;
+          }
+        }
```
Wouldn't this be cleaner as `var isBase64 = dataHeaders.some((header) => header.toLowerCase() === "base64")`?

Bhpsngum replied:
arrow function syntax does not exist before ES6, so im doing this to compatible with ES5 as well

mjwwit replied:
You are correct, and the rest of the project is still using ES5 syntax (even though ES2015 is allowed, as the minimum Node.js version is 12). However, you could still use `var isBase64 = dataHeaders.some(function (header) { return header.toLowerCase() === "base64"; })`

Bhpsngum replied:
Sure, i could do that

mjwwit commented on lib/XMLHttpRequest.js:
```diff
@@ -391,7 +452,61 @@ function XMLHttpRequest(opts) {
+
+        var base64 = false;
+        // check if header part has base64
+        var data_headers = parts[0].split(";");
```
Let's use camelCase everywhere for consistency

Bhpsngum replied:
alright i will change this one too

mjwwit reviewed (changes requested):
Quite a lot of code here, which isn't surprising given the functionality that has been added. This is no problem, but it will take some time to review everything. To not keep you waiting I've reviewed it partially.
I apologize for the number of small comments. It's a messy project already and I'd rather not make it any worse by introducing (more) inconsistencies.

Bhpsngum committed:
resolve some issues

Bhpsngum committed:
little docfix for params

Bhpsngum commented:
> Quite a lot of code here, which isn't surprising given the functionality that has been added. This is no problem, but it will take some time to review everything. To not keep you waiting I've reviewed it partially. I apologize for the number of small comments. It's a messy project already and I'd rather not make it any worse by introducing (more) inconsistencies.

It's all fine, after all this project is quite old already, and considering the number of dependencies, need to be careful indeed.

mjwwit commented:
I'm very sorry, but my time is very limited at this moment. I hope I can find the time to finish the review this weekend. Please bear with me.

Bhpsngum commented:
it's all fine, just take your time

mjwwit commented on lib/XMLHttpRequest.js:
```diff
@@ -46,6 +72,12 @@ function XMLHttpRequest(opts) {
   var https = require('https');
 
+  var max_redirects = opts.maxRedirects;
+  if (max_redirects == null || isNaN(max_redirects)) max_redirects = 20;
```
I'd really like to stay away from having non-strict equality operators. There's lots of hidden quirks (like `[] == false`) that are easily overlooked and the syntax isn't immediately obvious.
For the same reason, we should be using `Number.isNaN` as just `isNaN` will also check if the value can be coerced to a number (which is probably not what you want).

Looking at your case I'd replace this with:
```suggestion
  if (typeof max_redirects !== 'number' || Number.isNaN(max_redirects)) max_redirects = 20;
```

Bhpsngum replied:
It's a really good approach, thanks for this!

mjwwit commented on lib/XMLHttpRequest.js:
```diff
@@ -37,7 +36,34 @@ XMLHttpRequest.XMLHttpRequest = XMLHttpRequest;
+    allowFileSystemResources: true,
+    maxRedirects: 20, // Chrome standard
+    origin: undefined
+  }
```
```suggestion
  };
```

Bhpsngum committed:
Updates to reflect new reviews

mjwwit commented on lib/XMLHttpRequest.js:
```diff
@@ -868,4 +1064,4 @@ function XMLHttpRequest(opts) {
     }
   };
-};
+};
```
You're missing a final newline here 😉

mjwwit commented on tests/test-data-uri.js:
```diff
@@ -0,0 +1,131 @@
+  }, 500);
+}
+
+startTest();
```
Another missing newline here

mjwwit commented on tests/test-disallow-fs-resources.js:
```diff
@@ -0,0 +1,36 @@
+  xhr = new XMLHttpRequest({ allowFileSystemResources: false });
+
+  xhr.onreadystatechange = function() {
+    if (this.readyState == 4) {
```
`===`

mjwwit commented on tests/test-disallow-fs-resources.js:
```diff
@@ -0,0 +1,36 @@
+xhr = new XMLHttpRequest({ allowFileSystemResources: false });
+
+xhr.onreadystatechange = function() {
+  if (this.readyState == 4) {
```
`===`

mjwwit commented on tests/test-max-redirects.js:
```diff
@@ -0,0 +1,47 @@
+    let xhr = new XMLHttpRequest({ maxRedirects: 10 });
+    xhr.open("GET", "http://localhost:8888/redirectingResource/10", false);
+    xhr.onreadystatechange = function() {
+    if (xhr.readyState == 4) {
```
`===`

mjwwit commented on tests/test-max-redirects.js:
```diff
@@ -0,0 +1,47 @@
+    let xhr = new XMLHttpRequest({ maxRedirects: 10 });
+    xhr.open("GET", "http://localhost:8888/redirectingResource/20", false);
+    xhr.onreadystatechange = function() {
+      if (xhr.readyState == 4) {
```
`===`

mjwwit commented on tests/test-max-redirects.js:
```diff
@@ -0,0 +1,47 @@
+  } finally {
+    serverProcess.kill('SIGINT');
+  }
+}, 100);
```
Missing newline

mjwwit commented on tests/test-unsafe-redirect.js:
```diff
@@ -0,0 +1,41 @@
+}).listen(8000);
+
+xhr.onreadystatechange = function() {
+  if (this.readyState == 4) {
```
`===`

mjwwit commented on tests/test-url-origin.js:
```diff
@@ -0,0 +1,47 @@
+    };
+    xhr.send();
+  } catch(e) {
+    
```
I'd at least put some log statement here as to not fail silently.

mjwwit commented on tests/test-url-origin.js:
```diff
@@ -0,0 +1,47 @@
+    let xhr = new XMLHttpRequest({ origin: "http://localhost:8888/text" });
+    xhr.open("GET", "", false);
+    xhr.onreadystatechange = function() {
+      if (xhr.readyState == 4) {
```
`===`

mjwwit commented on tests/test-url-origin.js:
```diff
@@ -0,0 +1,47 @@
+  } finally {
+    serverProcess.kill('SIGINT');
+  }
+}, 100);
```
Missing newline

mjwwit reviewed (changes requested):
Just minor comments, overall this looks very good. Thanks a bunch!
I probably missed some usages of `==` or `!=`. Also, could you add your tests to the `test` script in package.json?

Bhpsngum committed:
Update lib/XMLHttpRequest.js

Co-authored-by: Michael de Wit <mjwwit@users.noreply.github.com>

Bhpsngum committed:
Update lib/XMLHttpRequest.js

Co-authored-by: Michael de Wit <mjwwit@users.noreply.github.com>

Bhpsngum committed:
Change from error code to error messages + newlines

Bhpsngum committed:
Missing strict equality somewhere

Bhpsngum commented:
I've fixed the reviews, also due to the thing that we agreed to change from `ERR_INVALID_URL` to `Invalid URL`, i also changed other ones from code to message to unify with this:
- `ERR_UNSAFE_REDIRECT` -->`Unsafe redirect`
- `ERR_TOO_MANY_REDIRECTS` --> `Too many redirects`

Bhpsngum commented:
oh wow, i guess i'll just write a nodejs script to automatically run the test case, the current one looks... not that nice to check.

mjwwit commented on tests/test-redirect-301.js:
```diff
@@ -0,0 +1,41 @@
+}).listen(8000);
+
+xhr.onreadystatechange = function() {
+  if (this.readyState == 4) {
```
One more here

mjwwit commented on tests/test-redirect-308.js:
```diff
@@ -0,0 +1,41 @@
+}).listen(8000);
+
+xhr.onreadystatechange = function() {
+  if (this.readyState == 4) {
```
And here

Bhpsngum replied:
oops let me push the changes real quick

Bhpsngum replied:
oh yeah i just copy from other redirects testing code (302 - 307), let me update them as well

mjwwit commented:
> oh wow, i guess i'll just write a nodejs script to automatically run the test case, the current one looks... not that nice to check.

It may be a good time to introduce something like that, thanks. Eventually we'll probably need a decent test runner though.

Bhpsngum commented:
I might introduce some unit testing and code coverage frameworks like Istanbul + Mocha combination, but as for now, I'm just writing a simple node script. Might be useful in the future though.

Bhpsngum committed:
Create automated test script

Bhpsngum committed:
change identation of test runner

Bhpsngum committed:
add strict equality in redirect code checks

mjwwit commented:
I'll release later today or tomorrow, thanks again!

Bhpsngum commented:
Thanks for the update as well 😄
I think we also need to document the features prior to the updates (including some old options like `agent` or ssl-related properties), so people can track down what features are available

mjwwit commented:
Good idea, I'll make sure that's done before the release.
